module/VuFindSearch/src/VuFindSearch/Backend/Primo/RestConnector.php

```diff
@@ -0,0 +1,705 @@
+            ];
+            uasort(
+                $facets[$facet->name],
+                function ($a, $b) {
+                    // Put the selected facets (with null as value) on the top:
+                    return ($b ?? PHP_INT_MAX) <=> ($a ?? PHP_INT_MAX);
+                }
+            );
+        }
+
+        // Apparently there's no "did you mean" data in the response..
+
+        return [
+            'recordCount' => $totalhits,
+            'documents' => $items,
+            'facets' => $facets,
+            'didYouMean' => [],
+            'error' => $response->info->errorDetails->errorMessages[0] ?? [],
+        ];
+    }
+
+    /**
+     * Process highlighting tags of the record fields
+     *
+     * @param array     $record    Record data
+     * @param array     $params    Request params
+     * @param \StdClass $highlight Highlighting data
+     *
+     * @return void
+     */
+    protected function processHighlighting(array &$record, array $params, \StdClass $highlight): void
+    {
+        if (empty($params['highlight'])) {
+            return;
+        }
+
+        $startTag = $params['highlightStart'] ?? '';
+        $endTag = $params['highlightEnd'] ?? '';
+
+        $highlightFields = [
+            'title' => 'title',
+            'creator' => 'author',
+            'description' => 'description',
+        ];
+
+        $hilightDetails = [];
+
+        foreach ($highlightFields as $primoField => $field) {
+            if (
+                ($highlightValues = $highlight->$primoField ?? null)
+                && !empty($record[$field])
+            ) {
+                $match = implode(
+                    '|',
+                    array_map(
+                        function ($s) {
+                            return preg_quote($s, '/');
+                        },
+                        $highlightValues
+                    )
+                );
+                $hilightDetails[$field] = array_map(
+                    function ($s) use ($match, $startTag, $endTag) {
+                        return preg_replace("/(\b|-|–)($match)(\b|-|–)/", "$1$startTag$2$endTag$3", $s);
+                    },
+                    (array)$record[$field]
+                );
+            }
+        }
+
+        $record['highlightDetails'] = $hilightDetails;
+    }
+
+    /**
+     * Fix the description field by removing tags etc.
+     *
+     * @param string $description Description
+     *
+     * @return string
+     */
+    protected function processDescription($description)
+    {
+        // Sometimes the entire article is in the description, so just take a chunk
+        // from the beginning.
+        $description = trim(mb_substr($description, 0, 2500, 'UTF-8'));
+        // These may contain all kinds of metadata, and just stripping
+        // tags mushes it all together confusingly.
+        $description = str_replace('<P>', '<p>', $description);
+        $paragraphs = explode('<p>', $description);
```

Comment:
If this is a problem, it's not a new one.... but just curious: I notice we're stripping tags, but we're not decoding entities. Is it possible this results in bad output somewhere, or does the data never contain entities, or am I just overlooking some detail?

Reply:
If my memory serves, the data is not "entity-encoded", and decoding could cause things to go wrong. But my recollection is not perfect. At least I've not seen any issues as of yet, and don't recall any being reported from our users.